TELCODOCS-1410: Recommend creating new catalog source for Operator image (#64611)
openshift/openshift-docs: modules/cnf-topology-aware-lifecycle-manager-operator-update.adoc

          registryPoll:
            interval: 1h
----

> <1> The index image URL contains the desired Operator images. If the index images are always pushed to the same image name and tag, this change is not needed. If you need to update the Operator image index, create a separate catalog source policy with the new image index or registry poll interval update. You must also prepare a separate subscription policy for the desired Operators to reference the new catalog source.

Reviewer: It doesn't need to be a separate policy. In PGT, it can be another sourceFile with the same policy name but different CR name and spec. Similarly, subscriptions in PGT just need an update on its `source` field.

<1> The index image URL contains the desired Operator images. If the index images are always pushed to the same image name and tag, this change is not needed.
<2> Set how frequently the Operator Lifecycle Manager (OLM) polls the index image for new Operator versions with the `registryPoll.interval` field. This change is not needed if a new index image tag is always pushed for y-stream and z-stream Operator updates. The `registryPoll.interval` field can be set to a shorter interval to expedite the update, however shorter intervals increase computational load. To counteract this, you can restore `registryPoll.interval` to the default value once the update is complete.
<3> Last observed state of the catalog connection. The `READY` value ensures that the `CatalogSource` policy is ready, indicating that the index pod is pulled and is running. This way, {cgu-operator} upgrades the Operators based on up-to-date policy compliance states.
+

> [NOTE]

Reviewer 2: This is a lot in a single note. Plus normally we try to avoid putting procedures/steps/actions in notes.

I'd recommend trying to pull this out of the note and making it another substep below. 

It seems worthwhile to be its own step. It could be something as simple as "Wait for OLM to update the subscription status." And then follow with the details about the remediation (if the status doesn't update, or what the issue is that persists). Or there might be a better approach.

But if there are situations where they really can't move on to the next step if they are stuck in the state that this note is talking about, then I think a substep would be more appropriate. And just more user-friendly, not having a note that takes up an entire page.

I'll continue with my review as if it's staying in the note, but I'd recommend another set of eyes if you do switch it to a substep

Author: Thanks @bergerhoffer 
Yes I was unsure about this one. Apparently this issue occurs in .08% of Operator updates according to testing, so it's considered an edge case. It's something that should be fixed as an enhancement in future but there's no timeline. 

For a bug, I was hesitant to start rewriting the whole section to satisfy the edge case with the new workflow. And I think the ideal scenario is the older workflow. Otherwise, in the new workflow, users will be adding configurations every time they need to update an Operator, which won't be ideal as the PolicyGenTemplate they are adding to will become unwieldy IMO. I tried a "shorter" note (not by much). If you still feel this breaks too many rules, I think I'd need to break out a troubleshooting module perhaps.

====

> After a catalog source update, it takes time for OLM to update the subscription status. This can mean that the status of the subscription policy may continue to show as compliant when TALM decides whether remediation is needed. As a result the Operator specified in the subscription policy does not get upgraded. To mitigate this issue, the `lastObservedState: READY` specification reduces the time between the completion of the catalog source policy remediation and the update of the subscription status. 

Reviewer: :+1:

> After a catalog source update, it takes time for OLM to update the subscription status. This can mean that the status of the subscription policy may continue to show as compliant when TALM decides whether remediation is needed. As a result the Operator specified in the subscription policy does not get upgraded. To mitigate this issue, the `lastObservedState: READY` specification reduces the time between the completion of the catalog source policy remediation and the update of the subscription status. 

Reviewer 3: "it" can be a vague referent, so maybe be more specific, such as  

> After a catalog source update, the OLM might take some time to update the subscription status.

Reviewer 3: According to the SSG, [admonitions](https://redhat-documentation.github.io/supplementary-style-guide/#admonitions) should be short and concise, so maybe an optional step that includes a note might work?

Reviewer 3: May/might

![Screenshot from 2023-09-25 16-42-16](https://github.com/openshift/openshift-docs/assets/57954076/7ee879bb-730e-4eab-8d38-2f67b7e2ffb7)

> After a catalog source update, it takes time for OLM to update the subscription status. This can mean that the status of the subscription policy may continue to show as compliant when TALM decides whether remediation is needed. As a result the Operator specified in the subscription policy does not get upgraded. To mitigate this issue, the `lastObservedState: READY` specification reduces the time between the completion of the catalog source policy remediation and the update of the subscription status. 

Reviewer 2: * A few suggestions, feel free to ignore if you disagree with any of them
* It's also unclear to me what the last sentence is saying. Does the user take action on the "lastObservedState" field? Or are you saying that OCP is mitigating the issue with this field somehow? Or you're just reinforcing why you told the user to add this into the config above?

```suggestion
After a catalog source update, it takes time for OLM to update the subscription status. The status of the subscription policy might continue to show as compliant while TALM decides whether remediation is needed. As a result, the Operator specified in the subscription policy does not get upgraded. To mitigate this issue, the `lastObservedState: READY` specification reduces the time between the completion of the catalog source policy remediation and the update of the subscription status. 
```

> However, if the issue persists, you can add another catalog source configuration to the `PolicyGenTemplate` resource with the same specifications, except for the following changes: 

Reviewer 3: I'm unclear if the "same specifications" indicate as some default settings or something more specific that is described elsewhere?

Reviewer 3: Just wondering if the follow YAML guideline applies here?

![Screenshot from 2023-09-25 16-49-27](https://github.com/openshift/openshift-docs/assets/57954076/4a2b4033-04c6-478b-8846-829e032a4a35)

[Link](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#yaml-formatting-for-kubernetes-and-openshift-api-objects)

Author: Oh, I wasn't aware of the ellipsis issue. Do you know if this rule is enforced @bergerhoffer ? I got 772 hits for this in the repo...

Reviewer 2: I know we talked about it on slack already, but just posting here too that, yes, we do try to enforce it. But it does get missed a lot, so it's a mixed bag in the repo.

> However, if the issue persists, you can add another catalog source configuration to the `PolicyGenTemplate` resource with the same specifications, except for the following changes: 

Reviewer 2: Kind of going along with my question above about the "lastObservedState" and the term "mitigate"; this is the step that seems like it's mitigating if they user has to do a workaround here. Maybe that term is more appropriate here?

[source,yaml]
----
- fileName: DefaultCatsrc.yaml
  # Contains original catalog source configuration
  ...
- fileName: DefaultCatsrc.yaml 
      remediationAction: inform
      policyName: "operator-catsrc-policy"
      metadata:
        name: redhat-operators-v2 <1>
      spec:
        displayName: Red Hat Operators Catalog v2 <2>
        image: registry.example.com:5000/olredhat-operators:<version> <3>

> ...

Reviewer 2: Based on the lead-in sentence, I'm unsure if I should add the "updateStrategy" section from the original one, since it says with the same specifications. (I don't know if the "..." here is for that, or if it is for the "status" section.

----
<1> Update the name for the new configuration.
<2> Update the display name for the new configuration.
<3> Update the index image URL. This `fileName.spec.image` field overrides any configuration in the `DefaultCatsrc.yaml` file.

> Update the `Subscription` resource to point to the new configuration for Operators that require an update:

Reviewer 2: This is the first that I'm seeing a Subscription resource mentioned in the module (and assembly). Should there be some information on how to get/update this resource? The YAML doesn't have a name or namespace defined, which I think we would need to know?

Author: The subscriptions are in the same repo as the `defaultcatsrc.yaml` file. As this is about updating Operators, the subscriptions should all already be present, it's more of a Day 2 operation. It can be a subscription to any Operator but I've just included some sample values.

[source,yaml]
----
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
spec:
  source: redhat-operators-v2 <1>
----
<1> Enter the name of the additional catalog source configuration that you defined in the `PolicyGenTemplate` resource.
====

.. This update generates one policy, `du-upgrade-operator-catsrc-policy`, to update the `redhat-operators` catalog source with the new index images that contain the desired Operators images.

> +

Reviewer 2: Something is now throwing off the formatting/numbering of the rest of the procedure:

![image](https://github.com/openshift/openshift-docs/assets/7443887/ac844650-a2af-461a-882c-fd068c97eb2d)

A "+" is visible in the source, and the rest of this isn't indented, and the substeps start over at "a" after the below note

Author: waiting for build to see if it's fixed

[NOTE]
====
If you want to use the image pre-caching for Operators and there are Operators from a different catalog source other than `redhat-operators`,  you must perform the following tasks:

> * Prepare a separate catalog source policy with the new index image or registry poll interval update for the different catalog source. 

Reviewer 3: I've had to read "for the different catalog source." a few times. Can this be dropped or maybe specific:

> you must perform the following tasks for the different catalog source:

Differently configured?

Reviewer 2: This line is whitespace trimming only, not something that was modified in the PR

* Prepare a separate subscription policy for the desired Operators that are from the different catalog source.
====

> 

Reviewer 3: The formatting of sub-steps is off. Maybe a plus symbol is needed on line 99?
